File: Multivariate/chapter6.R
sales = read.csv("D:/knou/Multivariate/data/sales.csv")
head(sales)

# 기술통계랑
summary(sales[,c(2:8)])

# 상관계수행렬 구하기

install.packages("ggplot2")
install.packages("GGally")
install.packages("CCA")
library(ggplot2)
library(GGally)
library(CCA)

exam = sales[,c(2:5)]
perform = sales[,c(6:8)]

# X1~X4 이루어진 객체 exam의 산점도 및 상관계수 구함
ggpairs(exam)


# X1~X4 이루어진 객체 exam의 산점도 및 상관계수 구함
ggpairs(exam)

# Y1~Y3 이루어진 객체 exam의 산점도 및 상관계수 구함
dev.new()
ggpairs(perform)

# 두 변수군 객체 exam과 perform의 상관계수행렬를 구하는 명령어
matcor(exam, perform)

# 정준상관분석 실행
cano_result = cc(exam, perform)
names(cano_result)

# 정준상관계수 저장 => 제 1,2정준상관계수가 값이 높아 유의미한 의미를 갖고 있음
cano_result$cor

# 정준상관분석 계수 결과
round(cano_result$xcoef, 3)

round(cano_result$ycoef, 3)

# 정준상관점수 & 정준변수부하 점수
cano_result$scores

# 정분변수 산점도 그리기 
w1 = cano_result$scores$xscores[,1]
y1 = cano_result$scores$yscores[,1]
cor(w1, y1)

plot(w1, y1, pch=19)

w2 = cano_result$scores$xscores[,2]
y2 = cano_result$scores$yscores[,2]
cor(w2, y2)

plot(w2, y2, pch=19)

# 3번은 상관관계가 낮음
w3 = cano_result$scores$xscores[,3]
y3 = cano_result$scores$yscores[,3]
cor(w3, y3)

plot(w3, y3, pch=19)

File: Multivariate/chapter6.py
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.cross_decomposition import CCA
import matplotlib.pyplot as plt
import seaborn as sns

sales = pd.read_csv("D:/knou/Multivariate/data/sales.csv")
print(sales.head(3))

# 기술통계량 
sales_desc = sales.iloc[:,1:].describe()
print(sales_desc)

# 표준편차 보기 
print(sales_desc.loc['std'])

# 변수군 선택
xlist = ['X1', 'X2', 'X3', 'X4']
ylist = ['Y1', 'Y2', 'Y3']

exam = sales[xlist]
perform = sales[ylist]

print(exam.head())
print(perform.head())

# zscore 표준화 방법1
scaler = StandardScaler()
zexam = scaler.fit_transform(exam)
zperform = scaler.fit_transform(perform)

# z-score 표준화 방법 2
X_mc = (exam - exam.mean()) / exam.std()
Y_mc = (perform - perform.mean()) / perform.std()

print(X_mc.head())
print(Y_mc.head())

# 정준상관분석 
# cca 모델 생성 : 최대 3개의 쌍으로 정렬된 선형 조합을 찾겠다는 의미
ca = CCA(n_components=3)
# 학습단계(상관성이 높은 선형 조합 계산)
ca.fit(exam, perform)
# 좌표계로 변환
X_c, Y_c = ca.transform(exam, perform)

# 정준상관점수 데이터프레임 만들기
cc_res = pd.DataFrame({"CCX_1":X_c[:,0],
 "CCY_1":Y_c[:,0],
 "CCX_2":X_c[:,1],
 "CCY_2":Y_c[:,1],
 "CCX_3":X_c[:,2],
 "CCY_3":Y_c[:,2]})

# 정준상관점수 보기
print(cc_res.head())

# 제1 정준상관계수 : X의 첫 번째 정준성분과 Y의 첫 번째 정준성분 간의 상관계수를 구함
print(np.corrcoef(X_c[:,0],Y_c[:,0]))
# 제2 정준상관계수
print(np.corrcoef(X_c[:,1],Y_c[:,1])[0,1])
# 제3 정준상관계수
print(np.corrcoef(X_c[:,2],Y_c[:,2])[0,1])

# 제1정준상관계수 산점도
sns.scatterplot(x="CCX_1", y="CCY_1", data=cc_res)
plt.title('corr = %.4f' %  np.corrcoef(X_c[:,0], Y_c[:,0])[0,1])
#plt.show()

# 정준상관계수 : 변수 4개, 정준성분 3개
print(ca.x_weights_)

# 정준상관계수 : 변수 4개, 정준성분 3개
print(ca.y_weights_)

# 정준상관점수 구하기
# CCX_1 in <Py 6.4> from zvalue in <Py 6.3> and x_weights_
# -0.5578*0.356-0.644*0.1468-0.729*0.2457-0.926*0.889
# Out[75]: -1.2954453
# CCY_1 in <Py 6.4> from zvalue in <Py 6.3> and y_weights_
# -0.795*0.7246-1.0149*0.379-1.063*0.575
# Out[76]: -1.5719291

